File: go/frontend/sound.go
package frontend

import (
	"encoding/binary"
	"log"
	"math"
	"time"

	"github.com/hajimehoshi/ebiten/v2/audio"
)

const (
	sampleRate = 44100
	beepHz     = 440
	beepVolume = 0.1
)

// Sound is a square-wave beeper that plays while the sound timer is active.
type Sound struct {
	player *audio.Player // nil when no audio device is available
}

// NewSound opens the audio device, or runs silently when there is none.
func NewSound() *Sound {
	player, err := audio.NewContext(sampleRate).NewPlayerF32(&squareWave{})
	if err != nil {
		log.Printf("Sound disabled, could not open audio device: %v", err)
		return &Sound{}
	}
	player.SetBufferSize(50 * time.Millisecond)
	return &Sound{player: player}
}

func (s *Sound) SetPlaying(playing bool) {
	if s.player == nil || playing == s.player.IsPlaying() {
		return
	}
	if playing {
		s.player.Play()
	} else {
		s.player.Pause()
	}
}

// squareWave is an endless 440Hz square wave, as 32-bit float stereo PCM.
type squareWave struct {
	pos int
}

func (s *squareWave) Read(buf []byte) (int, error) {
	const halfPeriod = sampleRate / beepHz / 2
	const frameSize = 8 // 2 channels * 4 bytes
	n := len(buf) / frameSize * frameSize
	for i := 0; i < n; i += frameSize {
		sample := float32(beepVolume)
		if (s.pos/halfPeriod)%2 == 1 {
			sample = -sample
		}
		bits := math.Float32bits(sample)
		binary.LittleEndian.PutUint32(buf[i:], bits)
		binary.LittleEndian.PutUint32(buf[i+4:], bits)
		s.pos = (s.pos + 1) % (halfPeriod * 2)
	}
	return n, nil
}

File: rust/src/frontend/sound.rs
use std::num::NonZero;

use rodio::source::{Function, SignalGenerator, Source};
use rodio::{DeviceSinkBuilder, MixerDeviceSink, Player};

const SAMPLE_RATE: u32 = 44100;
const BEEP_HZ: f32 = 440.0;

/// Square-wave beeper that plays while the sound timer is active.
pub struct Sound {
    // the device sink must stay alive for the player to keep producing sound;
    // None when no audio device is available
    output: Option<(MixerDeviceSink, Player)>,
}

impl Sound {
    /// Opens the default audio device, or runs silently when there is none.
    pub fn new() -> Self {
        let output = Self::open()
            .inspect_err(|e| eprintln!("Sound disabled, could not open audio device: {e}"))
            .ok();
        Sound { output }
    }

    fn open() -> Result<(MixerDeviceSink, Player), rodio::DeviceSinkError> {
        let sink = DeviceSinkBuilder::open_default_sink()?;
        let player = Player::connect_new(sink.mixer());
        let sample_rate = NonZero::new(SAMPLE_RATE).unwrap();
        player.append(SignalGenerator::new(sample_rate, BEEP_HZ, Function::Square).amplify(0.1));
        player.pause();
        Ok((sink, player))
    }

    pub fn set_playing(&self, playing: bool) {
        if let Some((_, player)) = &self.output {
            if playing {
                player.play();
            } else {
                player.pause();
            }
        }
    }
}
